src/couch_replicator_httpc_pool.erl

```diff
@@ -55,114 +58,113 @@ release_worker(Pool, Worker) ->
-        busy = Busy,
-        free = Free,
-        waiting = Waiting,
-        callers = Callers
-    } = State,
-    NewCallers0 = demonitor_client(Callers, Pid),
-    case Free -- [Pid] of
-    Free ->
-        case Busy -- [Pid] of
-        Busy ->
-            {noreply, State#state{callers = NewCallers0}};
-        Busy2 ->
-            case queue:out(Waiting) of
-            {empty, _} ->
-                {noreply, State#state{busy = Busy2, callers = NewCallers0}};
-            {{value, From}, Waiting2} ->
-                {ok, Worker} = ibrowse:spawn_link_worker_process(Url),
-                NewCallers1 = monitor_client(NewCallers0, Worker, From),
-                gen_server:reply(From, {ok, Worker}),
-                NewState = State#state{
-                    busy = [Worker | Busy2],
-                    waiting = Waiting2,
-                    callers = NewCallers1
-                },
-                {noreply, NewState}
-            end
-        end;
-    Free2 ->
-        {noreply, State#state{free = Free2, callers = NewCallers0}}
-    end;
+handle_cast({release_worker, Worker}, State0) ->
+    State1 = release_worker_int(Worker, State0),
+    State2 = maybe_supply_worker(State1),
+    {noreply, State2}.
 
-handle_info({'DOWN', Ref, process, _, _}, #state{callers = Callers} = State) ->
+
+handle_info({'EXIT', Worker, _Reason}, State0) ->
+    State1 = release_worker_int(Worker, State0),
+    State2 = maybe_supply_worker(State1),
+    {noreply, State2};
+
+handle_info({'DOWN', Ref, process, _, _}, #state{callers = Callers} = State0) ->
     case lists:keysearch(Ref, 2, Callers) of
         {value, {Worker, Ref}} ->
-            handle_cast({release_worker, Worker}, State);
+            State1 = release_worker_int(Worker, State0),
+            State2 = maybe_supply_worker(State1),
+            {noreply, State2};
         false ->
-            {noreply, State}
-    end.
+            {noreply, State0}
+    end;
+
+handle_info(refresh_dns, State0) ->
+    #state{url=Url, dns_data=Data0}=State0,
+    State1 = case couch_replicator_dns:lookup(Url) of
+        {ok, {Data0, TTL}} ->
+            State0#state{dns_ttl=TTL};
+        {ok, {Data1, TTL}} ->
+            #state{free=Free, busy=Busy, closing=Closing} = State0,
+            lists:foreach(fun ibrowse_http_client:stop/1, Free),
+            State0#state{
+                dns_data=Data1,
+                dns_ttl=TTL,
+                closing=Busy ++ Closing,
+                free=[],
+                busy=[]
+            };
+        {error, Error} ->
+            twig:log(
+                warn,
+                "couch_replicator_httpc_pool failed DNS lookup: ~p",
+                [Error]
+            ),
+            State0
+    end,
+    timer:send_after(State1#state.dns_ttl * 1000, refresh_dns),
+    {noreply, State1}.
+
+
+code_change(1, OldState, _Extra) ->
+    {state, Url, Limit, Free, Busy, Waiting, Callers} = OldState,
+    State0 = case couch_replicator_dns:lookup(Url) of
```

Comment:
This piece is way too similar to `handle_info(refresh_dns, State)` clause. It also quite similar to the piece of logic in `init/1`.
Would it be possible to abstract this functionality out into `refresh_dns` function?